### PR TITLE
CASSANDRA-19418  - Changes to report additional bulk analytics job stats for instrumentation

### PR DESCRIPTION
**Pull request by arjunashok**
## Changes

- Update to relay the results of the spark tasks to the executor to instrument the total no. of rows written.
- Abstracts the current `dialHome` implementation within `CassandraBulkWriterContext` to be pluggable via a `JobStats` interface, that can allow recording of job parameters from the spark driver.
- To allow reporting of stats such as no. rows, bytes written and resize detection, a `WriterResult` type is returned from each executor, and are reported by the driver after aggregation.
- The above abstraction also separates the recording of the metadata as the job progresses and separately publish them at the end of the job with success/failure status and failure reason where applicable.

## Testing
- Validated successful build run w/ unit/integration tests.

### CODE REVIEW

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/BulkWriterContext.java [hunk 1]
```diff
@@ -21,7 +21,9 @@
 
 import java.io.Serializable;
 
-public interface BulkWriterContext extends Serializable
+import org.apache.cassandra.spark.common.Reportable;
+
+public interface BulkWriterContext extends Serializable, Reportable
```
Comment (frankgh) [inline cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/BulkWriterContext.java]: I don't think this belongs in the `BulkWriterContext`. The context is something we share with executors. However the job stats are tracked from the driver only. I would avoid adding anything else to the BulkWriterContext

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/common/Reportable.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.cassandra.spark.common;
```
Comment (frankgh) [inline cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/common/Reportable.java]: should this be in a stats or metrics package instead?

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/common/Reportable.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+import java.util.Map;
+
+/**
+ * Interface to provide functionality to report Spark Job Statistics and/or properties
+ * that can optionally be instrumented. The default implementation merely logs these
+ * stats at the end of the job.
+ */
+public interface Reportable
```
Comment (frankgh) [inline cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/common/Reportable.java]: How about naming this class `BulkWriterJobStats` ?

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/RingInstance.java [hunk 4]
```diff
@@ -49,6 +49,7 @@ public RingInstance(ReplicaMetadata replica)
                          .datacenter(replica.datacenter())
                          .state(replica.state())
                          .status(replica.status())
+                         .token("")
```
Comment (frankgh) [inline cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/RingInstance.java]: why this change?

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/RingInstance.java [hunk 5]
```diff
@@ -125,40 +126,28 @@ private void writeObject(ObjectOutputStream out) throws IOException
         out.writeUTF(ringEntry.address());
         out.writeInt(ringEntry.port());
         out.writeUTF(ringEntry.datacenter());
-        out.writeUTF(ringEntry.load());
```
Comment (frankgh) [inline cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/RingInstance.java:128]: why do we remove these?

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/CassandraBulkSourceRelation.java [hunk 6]
```diff
@@ -107,17 +112,25 @@ private void persist(@NotNull JavaPairRDD<DecoratedKey, Object[]> sortedRDD, Str
...
         {
+            recordFailureStats(throwable.getMessage());
             LOGGER.error("Bulk Write Failed", throwable);
             throw new RuntimeException("Bulk Write to Cassandra has failed", throwable);
         }
         finally
         {
             try
             {
+                writerContext.publishJobStats();
```
Comment (frankgh) [inline cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/CassandraBulkSourceRelation.java]: I don't think this will do what you think it will do. This object _should_ be treated as a read-only object from the point of view of the executors. Data won't propagate back to the driver once the executors complete

**Review by frankgh — CHANGES_REQUESTED** [review]
I think there are some issues with the PR in the sense that the data "reported" from executors will never make it back to the driver. Maybe revisit that aspect and see if you are trying to do will work

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/common/Reportable.java (hunk 2 above)
Comment (arjunashok) [reply cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/common/Reportable.java]: Makes sense

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/common/Reportable.java (hunk 3 above)
Comment (arjunashok) [reply cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/common/Reportable.java]: This is not meant to be specific to the writer, so can be renamed to `JobStats` maybe?

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/RingInstance.java (hunk 5 above)
Comment (arjunashok) [reply cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/RingInstance.java:128]: Since We are now returning the `StreamResult` back from the tasks, the existing implementation will result in NPEs while serializing the contained `RingInstance`, due to many of these fields not being defined when we create the `RingInstance` from `ReplicaMetadata`.  The change removes the fields not being used from RingInstance from the serialization context.

Likewise, the change is also explicitly setting the `token` field to a default for the same reason, since `token` is part of the equals/hashcode validations.

Stack trace:

```
Caused by: org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 1.0 failed 4 times, most recent failure: Lost task 0.3 in stage 1.0 (TID 40) (172.20.10.7 executor driver): com.esotericsoftware.kryo.KryoException: Error during Java serialization.
Serialization trace:
instance (org.apache.cassandra.spark.bulkwriter.CommitResult)
commitResults (org.apache.cassandra.spark.bulkwriter.StreamResult)
	at org.apache.cassandra.spark.bulkwriter.util.SbwJavaSerializer.write(SbwJavaSerializer.java:58)
	at com.esotericsoftware.kryo.Kryo.writeObject(Kryo.java:575)
```

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/RingInstance.java (hunk 4 above)
Comment (arjunashok) [reply cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/RingInstance.java]: Answered in the response below

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/BulkWriterContext.java (hunk 1 above)
Comment (arjunashok) [reply cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/BulkWriterContext.java]: So, the functionality provided by the new interface is replacing the existing `dialHome` method in the `CassandraBulkWriterContext` .  

The thinking is that this is tied to the "context" that is shared across executors, as we "record" initial stats and job status stats at the executor level and the "inflight" stats at the task level.

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/CassandraBulkSourceRelation.java (hunk 6 above)
Comment (arjunashok) [reply cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/CassandraBulkSourceRelation.java]: So, the change is not propagating data back to the driver, but publishes stats from the executors. I am assuming here that the context is made available to the executors so what you are saying does not need to happen. Let me know if that makes sense.

I have been able to validate this using the in-jvm-dtests.

Comment (arjunashok) [reply cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/CassandraBulkSourceRelation.java]: You're right. Fixed to only report/publish stats from the driver.

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/BulkWriterContext.java (hunk 1 above)
Comment (arjunashok) [reply cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/BulkWriterContext.java]: We could potentially have this implemented by the concrete `CassandraBulkWriterContext` to keep it consistent with how `dialHome` is currently implemented.

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/common/JobStats.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+package org.apache.cassandra.spark.common;
+
+import java.util.Map;
+
+/**
+ * Interface to provide functionality to report Spark Job Statistics and/or properties
+ * that can optionally be instrumented. The default implementation merely logs these
+ * stats at the end of the job.
+ */
+public interface JobStats
```
Comment (frankgh) [inline cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/common/JobStats.java]: Since `JobStats` is not part of the context, we need to make sure we define this class as serializable, since the `BulkWriterContext` is a `Broadcast` variable

```suggestion
public interface JobStats extends Serializable
```

File: cassandra-analytics-core/src/test/java/org/apache/cassandra/spark/bulkwriter/MockBulkWriterContext.java [hunk 8]
```diff
@@ -77,6 +78,23 @@ public class MockBulkWriterContext implements BulkWriterContext, ClusterInfo, Jo
     private int sstableDataSizeInMB = 128;
     private int sstableWriteBatchSize = 2;
 
+    private final Map<String, String> jobStats = new HashMap<>();
+
+    public Map<String, String> jobStats()
+    {
+        return jobStats;
+    }
```
Comment (frankgh) [inline cassandra-analytics-core/src/test/java/org/apache/cassandra/spark/bulkwriter/MockBulkWriterContext.java]: NIT: missing line break
```suggestion
    }

```

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/WriteResult.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.cassandra.spark.bulkwriter;
+
+import java.util.List;
+
+import scala.Serializable;
```
Comment (frankgh) [inline cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/WriteResult.java]: shouldn't this be a `java.io.Serializable` instead?

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/WriteResult.java [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+public class WriteResult implements Serializable
```
Comment (frankgh) [inline cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/WriteResult.java:29]: let's add javadocs here

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/WriteResult.java [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    public final List<StreamResult> streamResults;
+    public final boolean clusterResizeDetected;
+
+    public WriteResult(List<StreamResult> streamResults, boolean clusterResizeDetected)
+    {
+        this.streamResults = streamResults;
+        this.clusterResizeDetected = clusterResizeDetected;
+    }
+
+    public List<StreamResult> streamResults()
```
Comment (frankgh) [inline cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/WriteResult.java:46]: javadocs

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/SSTableWriter.java [hunk 12]
```diff
@@ -59,6 +59,9 @@ public class SSTableWriter
     private final Map<Path, Digest> fileDigestMap = new HashMap<>();
     private final DigestAlgorithm digestAlgorithm;
 
+    private long rowCount = 0;
+
```
Comment (frankgh) [inline cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/SSTableWriter.java:63]: NIT: remove extra line break
```suggestion
```

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/SSTableWriter.java [hunk 13]
```diff
@@ -100,6 +103,15 @@ public void addRow(BigInteger token, Map<String, Object> boundValues) throws IOE
         }
         maxToken = token;
         cqlSSTableWriter.addRow(boundValues);
+        rowCount += 1;
+    }
+
+    /**
+     * @return the total number of rows added
```
Comment (frankgh) [inline cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/SSTableWriter.java]:
```suggestion
     * @return the total number of rows written
```

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/RecordWriter.java [hunk 14]
```diff
@@ -75,6 +75,8 @@ public class RecordWriter implements Serializable
     private SSTableWriter sstableWriter = null;
     private int outputSequence = 0; // sub-folder for possible subrange splits
 
+    private static final String STATS_KEY_RESIZE_DETECTED = "clusterResizeDetected";
```
Comment (frankgh) [inline cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/RecordWriter.java]: do we need this? it seem unused
```suggestion
```

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/CassandraBulkWriterContext.java [hunk 15]
```diff
@@ -56,10 +57,12 @@ public class CassandraBulkWriterContext implements BulkWriterContext, KryoSerial
     private final CassandraClusterInfo clusterInfo;
     private final SchemaInfo schemaInfo;
 
-    private CassandraBulkWriterContext(@NotNull BulkSparkConf conf,
-                                       @NotNull CassandraClusterInfo clusterInfo,
-                                       @NotNull StructType dfSchema,
-                                       SparkContext sparkContext)
+    private final Map<String, String> jobStats = new HashMap<>();
```
Comment (frankgh) [inline cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/CassandraBulkWriterContext.java]: should this be transient? I don't think we want to serialize this value to the executors?

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/RingInstance.java (hunk 5 above)
Comment (frankgh) [reply cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/RingInstance.java:128]: shouldn't we handle nulls during serialization/deserialization instead?

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/CassandraBulkSourceRelation.java [hunk 16]
```diff
@@ -129,6 +143,64 @@ private void persist(@NotNull JavaPairRDD<DecoratedKey, Object[]> sortedRDD, Str
...
+        List<StreamResult> streamResults = writeResults.stream()
+                                                       .map(WriteResult::streamResults)
+                                                       .flatMap(Collection::stream)
+                                                       .collect(Collectors.toList());
+
+        long rowCount = streamResults.stream().mapToLong(res -> res.rowCount).sum();
+        long totalBytesWritten = streamResults.stream().mapToLong(res -> res.bytesWritten).sum();
+        boolean hasClusterTopologyChanged = writeResults.stream()
+                                                        .map(WriteResult::isClusterResizeDetected)
+                                                        .anyMatch(b -> b);
```
Comment (frankgh) [inline cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/CassandraBulkSourceRelation.java]: NIT:
```suggestion
                                                        .anyMatch(WriteResult::isClusterResizeDetected);
```

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/CassandraBulkSourceRelation.java [hunk 17]
```diff
@@ -129,6 +143,64 @@ private void persist(@NotNull JavaPairRDD<DecoratedKey, Object[]> sortedRDD, Str
...
+        LOGGER.info("Bulk writer has written {} rows and {} bytes with cluster-resize status: {}",
```
Comment (frankgh) [inline cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/CassandraBulkSourceRelation.java]: NIT:
```suggestion
        LOGGER.info("Bulk writer job complete. rows={} bytes={} cluster_resize={}",
```

**Review by frankgh — COMMENTED** [review]
Thanks for addressing the comments. I have a few follow up comments

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/CassandraBulkSourceRelation.java [hunk 18]
```diff
@@ -129,6 +143,64 @@ private void persist(@NotNull JavaPairRDD<DecoratedKey, Object[]> sortedRDD, Str
...
+    }
+
+    /**
+     * Get a ref copy of BulkWriterContext broadcast variable and compose a function to transform a partition into StreamResult
+     *
+     * @param ctx BulkWriterContext broadcast variable
+     * @return FlatMapFunction
+     */
+    private static FlatMapFunction<Iterator<Tuple2<DecoratedKey, Object[]>>, WriteResult>
+    partitionsFlatMapFunc(Broadcast<BulkWriterContext> ctx, String[] columnNames)
```
Comment (JeetKunDoug) [inline cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/CassandraBulkSourceRelation.java]: Why rename this function from `writeRowsInPartition` to `partitionsFlatMapFunc` - the first is a descriptive name for what the function does, the second is a descriptive name for what interface the returned value implements... one is "business domain" and the other is "implementation detail", and I think the former has much more value than the latter.

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/BulkWriterContext.java [hunk 19]
```diff
@@ -21,7 +21,9 @@
 
 import java.io.Serializable;
 
-public interface BulkWriterContext extends Serializable
+import org.apache.cassandra.spark.common.JobStats;
+
+public interface BulkWriterContext extends Serializable, JobStats
```
Comment (JeetKunDoug) [inline cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/BulkWriterContext.java]: Rather than extending JobStats, please add a `jobStats` method to this interface and return an instance of something that implements `JobStats` like the other methods here do. We did this originally because the number of things this interface included was huge, so we decided to break them up into smaller implementation classes.

An example `JobStatsImpl` class might look something like this:
```java
package org.apache.cassandra.spark.bulkwriter;

import java.util.HashMap;
import java.util.Map;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import org.apache.cassandra.spark.common.JobStats;

class JobStatsImpl implements JobStats
{
    private static final Logger LOGGER = LoggerFactory.getLogger(JobStatsImpl.class);;
    Map<String, String> stats = new HashMap<>();

    public JobStatsImpl(String sparkVersion, JobInfo jobInfo)
    {
        stats.put("sparkVersion", sparkVersion);
        stats.put("jobId", jobInfo.getId().toString());
        stats.put("keyspace", jobInfo.getId().toString());
        stats.put("table", jobInfo.getId().toString());
    }

    public void record(Map<String, String> statsToRecord)
    {
        stats.putAll(statsToRecord);
    }

    public void publish()
    {
        LOGGER.info("Job Stats: {}", stats);
    }
}
```

Then, CassandraBulkWriterContext just creates a new one of these in its constructor and returns it from the `jobStats` method. The actual field can/should be marked `transient` as Francisco has pointed out, so it doesn't get transferred to executors (and, therefore, it _doesn't_ need to be Serializable).

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/RingInstance.java (hunk 5 above)
Comment (JeetKunDoug) [reply cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/RingInstance.java:128]: Either that, or if the fields really aren't used, can we simply remove them completely? But, to Francisco's point, we really should handle this cleanly otherwise.

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/common/JobStats.java (hunk 7 above)
Comment (JeetKunDoug) [reply cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/common/JobStats.java]: See my other comments about how to implement this in BulkWriterContext - may not need to be Serializable if we move it to an instance variable like we've done with other parts of the BulkWriterContext interface.

**Review by JeetKunDoug — CHANGES_REQUESTED** [review]
Generally agree with Francisco's comments - please address them and let us know when you're ready for a second review round. I've added a few other requests here, but generally looks reasonable.

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/CassandraBulkSourceRelation.java (hunk 18 above)
Comment (arjunashok) [reply cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/CassandraBulkSourceRelation.java]: Makes sense. Reverted

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/RingInstance.java (hunk 5 above)
Comment (arjunashok) [reply cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/RingInstance.java:128]: Unfortunately, the fields are members of the `RingEntry` instance (part of the sidecar client library) so they cannot be removed here.  

I can add checks for the ones not in `ReplicaMetadata` for completeness, but the intention here was to not write the fields that we know are not used anyway.

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/WriteResult.java (hunk 9 above)
Comment (arjunashok) [reply cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/WriteResult.java]: Correct. fixed.

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/CassandraBulkWriterContext.java (hunk 15 above)
Comment (arjunashok) [reply cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/CassandraBulkWriterContext.java]: Addressed

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/BulkWriterContext.java (hunk 19 above)
Comment (arjunashok) [reply cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/BulkWriterContext.java]: Makes sense. Incorporated this change

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/common/stats/JobStatsImpl.java [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+public class JobStatsImpl implements JobStats
+{
+    private static final Logger LOGGER = LoggerFactory.getLogger(JobStatsImpl.class);
+    private final transient Map<String, String> jobStats = new HashMap<>();
+    public void recordJobStats(Map<String, String> stats)
+    {
+        jobStats.putAll(stats);
+    }
+
+    public void publishJobStats()
```
Comment (frankgh) [inline cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/common/stats/JobStatsImpl.java]: can we add override annotations?

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/RingInstance.java [hunk 21]
```diff
@@ -130,6 +130,22 @@ private void writeObject(ObjectOutputStream out) throws IOException
         out.writeUTF(ringEntry.fqdn());
         out.writeUTF(ringEntry.status());
         out.writeUTF(ringEntry.state());
+        if (ringEntry.rack() != null)
```
Comment (frankgh) [inline cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/RingInstance.java]: I think this is not sufficient, I pinged you offline for a suggestion

Comment (arjunashok) [reply cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/RingInstance.java]: Addressed

**Review by frankgh — APPROVED** [review]
+1 Thanks for addressing the comments
